### DISCUSSION TITLE
`onError` and `setErrorHandler`

### DISCUSSION THREAD

**Discussion by kraftwerk28** [fastify/fastify#3881] [Q&A] [answered]
Consider the following example:

```js
import fastify from "fastify";
import fp from "fastify-plugin";
import knex from "knex";

const app = fastify({
  logger: true,
});

const knexPlugin = fp(async (app, { connectionString }) => {
  const k = knex({
    client: "pg",
    connection: { connectionString },
  });
  app.decorate("knex", k);
  app.decorateRequest("tx", {
    async getter() {
      return (this.pendingTx = await app.knex.transaction());
    }
  });
  app.addHook("onError", async (req, res, err) => {
    // NOT triggered unless we call `setErrorHandler` below.
    if (req.pendingTx) {
      app.log.info("Rollbacking pending tx");
      await req.pendingTx.rollback();
      delete req.pendingTx;
    }
  });
  app.addHook("onResponse", async (req) => {
    if (req.pendingTx) {
      req.log.info("Committing pending tx");
      await req.pendingTx.commit();
      delete req.pendingTx;
    }
  });
});

// If we uncomment the next 3 lines, the `onError` hook starts working
// app.setErrorHandler(async (err) => {
//   throw err;
// });

app.register(knexPlugin, {
  connectionString: process.env.PG_DATABASE_URL,
});

app.post(
  "/",
  async (req, res) => {
    const tx = await req.tx; // Trigger transaction creation
    throw { statusCode: 429 }; // Try to trigger ROLLBACK
  },
);

await app.listen(parseInt(process.env.PORT));
```

From the [`onError` hook docs](https://www.fastify.io/docs/latest/Reference/Hooks/#onerror):
> This hook will be executed only after the customErrorHandler has been executed, and only if the customErrorHandler sends an error back to the user (Note that the default customErrorHandler always sends the error back to the user).

According to it, we have the default `customErrorHandler`, when no `setErrorHandler` is executed. Why `onError` hooks aren't triggered then?

**Comment by metcoder95** [comment, marked as answer]
Which version of Fastify are you using? I cannot reproduce it with latest `main`

**Reply by kraftwerk28** [reply]
I was using `fastify@3.29.0`, but with `fastify@4.0.0-rc.2` I can't reproduce the issue as well. Thank you!

**Reply by kraftwerk28** [reply]
Is it possible to have this fix in `fastify@^3`? A lot of fastify plugins require this major and won't work with `fastify@^4`.

**Reply by mcollina** [reply]
Unfortunately is a major change. You can track which plugins have been ported to fastify v4 in https://github.com/fastify/fastify/issues/3482. We are actively working on it.
